warehouse/subscriptions/services.py

```diff
@@ -0,0 +1,664 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import random
+
+from string import ascii_letters, digits
+
+import stripe
+
+from sqlalchemy import or_
+from sqlalchemy.orm.exc import NoResultFound
+from zope.interface import implementer
+
+from warehouse.organizations.models import (
+    OrganizationStripeCustomer,
+    OrganizationStripeSubscription,
+)
+from warehouse.subscriptions.interfaces import IBillingService, ISubscriptionService
+from warehouse.subscriptions.models import (
+    StripeSubscription,
+    StripeSubscriptionItem,
+    StripeSubscriptionPrice,
+    StripeSubscriptionPriceInterval,
+    StripeSubscriptionProduct,
+    StripeSubscriptionStatus,
+)
+
+
+class GenericBillingService:
+    def __init__(self, api, publishable_key, webhook_secret):
+        self.api = api
+        self.publishable_key = publishable_key
+        self.webhook_secret = webhook_secret
+
+    @classmethod
+    def create_service(cls, context, request):
+        """
+        Create appropriate billing service based on environment
+        """
+        raise NotImplementedError
+
+    def get_checkout_session(self, session_id, **kwargs):
+        """
+        Fetch the Checkout Session to based on the session_id passed to the success page
+        """
+        checkout_session = self.api.checkout.Session.retrieve(
+            session_id,
+            expand=["customer", "line_items", "subscription"],
+        )
+        return checkout_session
+
+    def get_customer(self, subscription_id):
+        """
+        Fetch the Customer resource attached to the Subscription
+        """
+        subscription = self.api.Subscription.retrieve(
+            subscription_id,
+            expand=["customer"],
+        )
+        return subscription.customer
+
+    def create_customer(self, name, description):
+        """
+        Create the Customer resource via Billing API with the given name and description
+        """
+        return self.api.Customer.create(
+            name=name,
```

Comment:
I'd like for us to give the customers objects in Stripe names like "PyPI Organization - " since the account is shared between other PSF usage (namely pycon).

Reply:
Added customer name prefix in commit [f292504](https://github.com/pypi/warehouse/pull/11986/commits/f29250421bbae77d06abcc7bb7f4b59da6398f20).